src/erlcloud_ec2.erl

```diff
@@ -1941,46 +2051,187 @@ describe_snapshot_attribute(SnapshotID, create_volume_permission, Config)
 describe_snapshots(SnapshotIDs, Owner, RestorableBy, Config)
-  when is_list(SnapshotIDs),
-       is_list(Owner) orelse Owner =:= none,
-       is_list(RestorableBy) orelse RestorableBy =:= none ->
-    Params = [{"Owner", Owner}, {"RestorableBy", RestorableBy}|
-              erlcloud_aws:param_list(SnapshotIDs, "SnapshotId")],
+    when is_list(SnapshotIDs),
+         is_list(Owner) orelse Owner =:= none,
+         is_list(RestorableBy) orelse RestorableBy =:= none ->
+    Params = erlcloud_aws:param_list(SnapshotIDs, "SnapshotId") ++ [{"Owner", Owner}, {"RestorableBy", RestorableBy}],
     case ec2_query(Config, "DescribeSnapshots", Params) of
         {ok, Doc} ->
             {ok, [extract_snapshot(Item) || Item <- xmerl_xpath:string("/DescribeSnapshotsResponse/snapshotSet/item", Doc)]};
         {error, _} = Error ->
             Error
+    end;
+describe_snapshots(Owner, RestorableBy, MaxResults, Config)
+    when is_list(Owner), is_list(RestorableBy), is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX, is_record(Config, aws_config) ->
+    describe_snapshots(Owner, RestorableBy, MaxResults, undefined, Config);
+describe_snapshots(Owner, RestorableBy, MaxResults, NextToken)
+    when is_list(Owner), is_list(RestorableBy), is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX, is_list(NextToken) ->
+    describe_snapshots(Owner, RestorableBy, MaxResults, NextToken, default_config());
+describe_snapshots(Owner, MaxResults, NextToken, Config)
+    when is_list(Owner), is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX, is_list(NextToken), is_record(Config, aws_config) ->
+    describe_snapshots(Owner, none, MaxResults, NextToken, Config).
+
+-spec describe_snapshots(ec2_snapshot_owner(), ec2_snapshot_restorable_by(), ec2_max_result(), ec2_token(), aws_config()) -> ec2_paged_describe_return().
+describe_snapshots(Owner, RestorableBy, MaxResults, NextToken, Config)
+    when is_list(Owner) orelse Owner =:= none,
+         is_list(RestorableBy) orelse RestorableBy =:= none,
+         is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX,
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    Params = [{"Owner", Owner}, {"RestorableBy", RestorableBy}, {"MaxResults", MaxResults}, {"NextToken", NextToken}],
+    case ec2_query(Config, "DescribeSnapshots", Params, ?NEW_API_VERSION) of
+        {ok, Doc} ->
+            Snapshots = [extract_snapshot(Item) || Item <- xmerl_xpath:string("/DescribeSnapshotsResponse/snapshotSet/item", Doc)],
+            NTxml = xmerl_xpath:string("/DescribeSnapshotsResponse/nextToken", Doc),
+            NewNextToken =
+                case NTxml of
+                    [] ->
+                        undefined;
+                    NTxml ->
+                        element(5,lists:nth(1,element(9,lists:nth(1,NTxml))))
```

Comment:
It extracts the Next Token from the response. I've moved it to a separate function